extension/src/background/helpers/session.ts

```diff
@@ -17,3 +28,249 @@ export class SessionTimer {
+
+  if (hashKey?.key && hashKey?.iv) {
+    try {
+      const format = "jwk";
+      // JSON Web Key can be parsed with decoding
+      const exportedHashKey = JSON.parse(hashKey.key) as JsonWebKey;
+      // import the password key for future use indecryption
+      const key = await crypto.subtle.importKey(
+        format,
+        exportedHashKey,
+        TEMPORARY_STORE_ENCRYPTION_NAME,
+        true,
+        ["encrypt", "decrypt"],
+      );
+
+      return {
+        iv: decode(hashKey.iv),
+        key,
+      };
+    } catch (e) {
+      return null;
+    }
+  }
+
+  return null;
+};
+
+interface StoreEncryptedTemporaryData {
+  localStore: DataStorageAccess;
+  keyName: string;
+  temporaryData: string;
+  hashKey: {
+    key: CryptoKey;
+    iv: ArrayBuffer;
+  };
+}
+
+export const storeEncryptedTemporaryData = async ({
+  localStore,
+  keyName,
+  temporaryData,
+  hashKey,
+}: StoreEncryptedTemporaryData) => {
+  const encryptedPrivateKey = await encryptHashString({
+    str: temporaryData,
+    keyObject: hashKey,
+  });
+
+  const existingTemporaryStore = await localStore.getItem(TEMPORARY_STORE_ID);
+
+  // store encrypted private key in local storage, a separate space from where the password is stored
+  await localStore.setItem(TEMPORARY_STORE_ID, {
+    ...existingTemporaryStore,
+    [keyName]: encode(encryptedPrivateKey),
+  });
+};
+
+interface GetEncryptedTemporaryData {
+  sessionStore: Store;
+  localStore: DataStorageAccess;
+  keyName: string;
+}
+
+export const getEncryptedTemporaryData = async ({
+  sessionStore,
+  localStore,
+  keyName,
+}: GetEncryptedTemporaryData) => {
+  const temoraryStore = (await localStore.getItem(TEMPORARY_STORE_ID)) || {};
+  const encryptedKeyJSON = temoraryStore[keyName];
+  if (!encryptedKeyJSON) {
+    return "";
+  }
+  const encryptedKey = decode(encryptedKeyJSON as string);
+  const hashKey = hashKeySelector(sessionStore.getState() as SessionState);
+
+  if (!hashKey?.key || !hashKey.iv || !encryptedKey) {
+    return "";
+  }
+
+  if (hashKey) {
+    // JSON Web Key can be parsed with decoding
+    const exportedHashKey = JSON.parse(hashKey.key) as JsonWebKey;
+    // import the password key for future use indecryption
+    const key = await crypto.subtle.importKey(
+      "jwk",
+      exportedHashKey,
+      TEMPORARY_STORE_ENCRYPTION_NAME,
+      false,
+      ["encrypt", "decrypt"],
+    );
+
+    // use the hashed password to decrypt the private key
+    const activePrivateKey = await decryptHashString({
+      hash: encryptedKey,
+      keyObject: { iv: decode(hashKey.iv), key },
+    });
+
+    return activePrivateKey;
+  }
+
+  return "";
+};
+
+interface ClearSession {
+  localStore: DataStorageAccess;
+  sessionStore: Store;
+}
+
+export const clearSession = async ({
+  localStore,
+  sessionStore,
+}: ClearSession) => {
+  sessionStore.dispatch(timeoutAccountAccess());
+  await localStore.remove(TEMPORARY_STORE_ID);
```

Comment:
when we end sessions, we now also clear `TEMPORARY_STORE_ID` in addition to dispatching `timeoutAccountAccess`